File: Java/dp/length_longest_common_substring.java
/*Longest Common Substring
Problem statement: Given two string s1 and s2, find the length of longest common substring of these two strings.
                   Substring - Contegeous sequence of characters within a string.Note that the string are case sensitive.
                   For e.g hello and Hello will have only 4, length of longest common substring since H and h are different.
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class LengthOfSubstring{

    // function to return the length of longest common substring
   public static void  LongestCommonSubstring(){
        Scanner sc = new Scanner(System.in);
        
        int cnt = 0;    //cnt is a variable to store the length of common substring
        String s1 = sc.nextLine();                 //input the first string from user
        String s2 = sc.nextLine();                // input the second string from user

        int fin_tab[][] = new int[s1.length() + 1][s2.length() + 1];            // matrix or table to store the length of longest common suffixes of substrings
        // filling the matrix using dynamic programming
        for(int itr1 = 1; itr1 < fin_tab.length; itr1++){                  
            for(int itr2 = 1; itr2 < fin_tab[0].length; itr2++){
                char str1 = s1.charAt(itr1 - 1);
                char str2 = s2.charAt(itr2 - 1);

                if(str1 == str2){
                     fin_tab[itr1][itr2] = fin_tab[itr1-1][itr2-1] + 1;
                }
                else {
                    fin_tab[itr1][itr2] = 0;
                }
                if(fin_tab[itr1][itr2] > cnt){
                    cnt = fin_tab[itr1][itr2];
                }
            }
        }     
        System.out.print("Length is " + cnt);
    }   

	public static void main (String[] stringcom){
            LongestCommonSubstring();   //calling the function for the length
	}
}

/*
Test Case 1:
Input:
helloword
loperd
Output: Length is 2
      
Test Case 2:
Input:
Towerofhanoi
Ofhand
Output: Length is 4
      
Time Complexity: O(size1*size2)

Space Complexity: O(size1*size2)
*/

File: Python/dp/length_longest_common_substring.py
""" Length of Longest Common Substring problem in Python

Given two strings, we have to find the length of the longest common substring in Python

The problem can be solved using Dynamic Programming
"""


def lengthoflongestcommonsubstring(S1, S2):
    # Build a 2D matrix in Bottom-Up manner. Every element
    # dp[i][j] contains the length of the longest common substring
    # of S1[0 .. i] and S2[0 .. j].
    n = len(S1)
    m = len(S2)
    ans = 0
    # Fill every element of the 2D matrix with 0 as value
    dp = [[0 for i in range(m+1)] for j in range(n+1)]

    for i in range(0, n + 1):
        for j in range(0, m + 1):

            # length of LCS when one of the string is empty is zero
            if 0 in (i, j):
                dp[i][j] = 0

            # If the elements of both the strings match, then
            # the value of LCS increments by one.
            # Also Check if the value of lcs has reached the max till now.
            elif S1[i-1] == S2[j-1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
                ans = max(ans, dp[i][j])
            else:
                dp[i][j] = 0
    return ans


if __name__ == '__main__':
    s1 = input("Enter the first string: ")
    s2 = input("Enter the second string: ")
    res = lengthoflongestcommonsubstring(s1, s2)
    print("The length of the longest common substring is " + str(res))


# Time Complexity: O(n * m), where n, m are the size of the given strings
# Space Complexity: O(n * m)

# SAMPLE INPUT AND OUTPUT

# Enter the first string: runner
# Enter the second string: run
# The length of the longest common substring is 3
